### PR TITLE
[query] New Shuffler with Arbitrary Sort Order

### PR DESCRIPTION
**Pull request by johnc1231**
This PR:

- Introduces `StructComparisonOp`, a subclass of `ComparisonOp` that is just for comparing structs. They have an array of sort fields for this purpose.
- Extends `StreamDistribute` to take in a `ComparisonOp`, so that it doesn't just assume data sorted in ascending order.
- Updates `LowerDistributedSort.scala` to actually use the `SortFields` that get passed into it, allowing for arbitrary sort orders.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/ComparisonOp.scala [hunk 1]
```diff
@@ -89,4 +92,36 @@ case class Compare(t1: Type, t2: Type) extends ComparisonOp[Int] {
...
+
+  override def codeOrdering(ecb: EmitClassBuilder[_], t1: SType, t2: SType): F[op.ReturnType] = {
+    ComparisonOp.checkCompatible(t1.virtualType, t2.virtualType)
+    ecb.getStructOrderingFunction(t1.asInstanceOf[SBaseStruct], t2.asInstanceOf[SBaseStruct], sortFields, op).asInstanceOf[CodeOrdering.F[op.ReturnType]]
+  }
+}
+
+case class StructCompare(t1: Type, t2: Type, sortFields: Array[SortField]) extends StructComparisonOp[Int] {
+  val op: CodeOrdering.Op = CodeOrdering.StructCompare()
+  override val strict: Boolean = false
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/expr/ir/ComparisonOp.scala:108]: Admittedly am not 100% sure what `strict` means in this context.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/ComparisonOp.scala:108]: strict refers to missing semantics -- strict=true means the result is missing if any argument is missing.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/ComparisonOp.scala:108]: I think this all looks correct.

**Review by tpoterba — APPROVED** [review]
🚀

**Pull request merged**